openupgrade_scripts/scripts/account/17.0.1.2/end-migration.py

```diff
@@ -0,0 +1,107 @@
+        for xmlid, journal_data in list(
+            template_data.get("account.journal", {}).items()
+        ):
+            if not env.ref(xmlid, raise_if_not_found=False):
+                journal = None
+                lang = Template._get_untranslatable_fields_target_language(
+                    chart_template, company
+                )
+                translated_code = Template._get_field_translation(
+                    journal_data, "code", lang
+                )
+                if "code" in journal_data:
+                    journal_code = translated_code or journal_data["code"]
+                    journal = (
+                        env["account.journal"]
+                        .with_context(active_test=False)
+                        .search(
+                            [
+                                *env["account.journal"]._check_company_domain(
+                                    company_id
+                                ),
+                                ("code", "=", journal_code),
+                            ]
+                        )
+                    )
+                # Try to match by journal name to avoid conflict in the unique
+                # constraint on the mail alias
+                translated_name = Template._get_field_translation(
+                    journal_data, "name", lang
+                )
+                if not journal and "name" in journal_data and "type" in journal_data:
+                    journal = (
+                        env["account.journal"]
+                        .with_context(active_test=False)
+                        .search(
+                            [
+                                *env["account.journal"]._check_company_domain(
+                                    company_id
+                                ),
+                                ("type", "=", journal_data["type"]),
+                                ("name", "in", (journal_data["name"], translated_name)),
+                            ],
+                            limit=1,
+                        )
+                    )
+                # That part has been added after Odoo code
+                # If we have not found a journal yet we search only by type making
+                # sure we do not assign again an already assigned journal
+                if not journal and "type" in journal_data:
+                    journal = (
+                        env["account.journal"]
+                        .with_context(active_test=False)
+                        .search(
+                            [
+                                *env["account.journal"]._check_company_domain(company),
+                                ("type", "=", journal_data["type"]),
+                                ("id", "not in", assigned_journals.ids),
+                            ],
+                            limit=1,
+                        )
+                    )
```

Comment:
maybe you could do two calls of this loop:

```python
for xmlid, journal_data in list(
            template_data.get("account.journal", {}).items()
        ):
```
the first loop containing the exact odoo code. and the second loop only containing this part you are adding if xmlid doesn't exist yet.